lib/src/components/NxCombobox/NxCombobox.tsx

```diff
@@ -0,0 +1,311 @@
+      });
+
+  // There is a requirement that when there is an error querying the data, if the user navigates away from
+  // the component and then comes back to it the search should be retried automatically
+  function handleComponentFocus(evt: FocusEvent<HTMLDivElement>) {
+
+    if (loadError) {
+      // check if this is focus coming into the component from somewhere else on the page, not just moving between
+      // children of this component and not from focus coming into the browser from some other window
+      const comingFromOutsidePage = evt.relatedTarget === null,
+          comingFromChildNode = evt.relatedTarget instanceof Node && evt.currentTarget.contains(evt.relatedTarget);
+
+      if (!(comingFromOutsidePage || comingFromChildNode)) {
+        doSearch(value);
+      }
+    }
+  }
+
+  function handleComponentBlur(evt: FocusEvent<HTMLDivElement>) {
+
+    if (!(evt.relatedTarget instanceof Node && evt.currentTarget.contains(evt.relatedTarget))) {
+      // The automatically selected suggestion becomes the value of the combobox
+      // when the combobox loses focus.
+      if (autoComplete && focusableBtnIndex !== null) {
+        const elToFocusText = matches[focusableBtnIndex].displayName;
+        onChange(elToFocusText);
+        onSearch(elToFocusText.trim());
+      }
+      setFocusableBtnIndex(null);
+      setInlineStyle(false);
+    }
+  }
+
+  function getDropdownBtnIdForIndex(idx: number) {
+    return `${dropdownBtnIdPrefix}-${idx}`;
+  }
+
+  function handleOnChange(newVal: string) {
+    setFocusableBtnIndex(null);
+    onChange(newVal);
+
+    if (newVal.trim() !== value.trim()) {
+      doSearch(newVal);
+    }
+  }
+
+  function doSearch(val: string) {
+    focusTextInput();
+    onSearch(val.trim());
+  }
+
+  function focusTextInput() {
+    inputRef.current?.querySelector('input')?.focus();
+  }
+
+  // helper for focusing different buttons in the dropdown menu
+  const adjustBtnFocus = (adjust: (i: number) => number) => () => {
+        const newFocusableBtnIndex = adjust(focusableBtnIndex ?? 0),
+            elToFocus = dropdownRef.current?.children[newFocusableBtnIndex] as HTMLElement | null;
+
+        if (elToFocus) {
+          setFocusableBtnIndex(newFocusableBtnIndex);
+          elToFocus.scrollIntoView({ block: 'nearest' });
+          setInlineStyle(false);
+        }
+      },
+      focusNext = adjustBtnFocus(inc),
+      focusPrev = adjustBtnFocus(dec),
+      focusFirst = adjustBtnFocus(always(0)),
+      focusLast = adjustBtnFocus(always(matches.length - 1));
+
+  function handleOnClick({displayName}: DataItem<T, string>) {
+    onChange(displayName);
+    onSearch(displayName.trim());
+    focusTextInput();
+    setFocusableBtnIndex(null);
+    setInlineStyle(false);
+  }
+
+  function handleKeyUp(evt: KeyboardEvent<HTMLElement>) {
+    if (autoComplete) {
+      const isPrintableChar = evt.key.length === 1 && evt.key.match(/[!-~]/);
+      setInlineStyle(!!isPrintableChar);
+    }
+  }
```

Comment:
Let's move these two up a but so that `adjustBtnFocus` and its derivatives are next to the main place they get used: `handleKeyDown`

Reply:
Updated here 5bcbfc054af23a527b539b8a606aac3bf575fa22.